Adds a sanity check to PriusVis.

liangfok opened the pull request:
Adds a check to PriusVis insisting that the RigidBodyTree only contain a single model instance that's connected to the world via an RPY joint. This was originally motivated by https://github.com/RobotLocomotion/drake/pull/5635#issuecomment-289596389.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5638)
<!-- Reviewable:end -->

liangfok commented:
`bazel-debug` being evaluated here: https://drake-jenkins.csail.mit.edu/view/Experimental/job/linux-xenial-clang-bazel-experimental/3/

liangfok commented:
New test is at: https://drake-jenkins.csail.mit.edu/view/Experimental/job/linux-xenial-clang-bazel-experimental/4/

liangfok commented:
+@seancurtis-tri for feature review please.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-:-KgHUCYrDei-_oZIaQN8:b-4269ug)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Generally, this feels a round-about way of accomplishing the task.  I'd recommend the following:

1. Call `std::vector<int> RigidBodyTree::FindBaseBodies()`
2. Assert that the size of the vector is 1. (A single body attached to the world implies a single model instance)
3. Grab the body for the single "root" body (e.g., `RigidBody<T>& get_body(int body_index) const`
4. Attempt to cast the joint to RPY (via `RigidBody::getJoint() const`)

This feels like it would more directly and unambiguously confirm the documented invariant -- a single model instance attached by a RPY joint.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/prius_vis.cc, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgKFmasAVjuEbU3OV9n:-KgKFmasAVjuEbU3OV9o:b-o9e4jv) ([raw file](https://github.com/robotlocomotion/drake/blob/565df173b28e825767020ce573ac6a7692e61b01/drake/automotive/prius_vis.cc#L50)):*
> ```cpp
>     const Isometry3<T>& X_WM) const {
>   // This method requires that tree_ contains only a single model instance and
>   // that it is connected to the world via a RollPitchYawFloatingJoint.
> ```

I would think this comment belongs in the documentation of this method's interface.  It should go somewhere in the header file.

---

*[drake/automotive/prius_vis.cc, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgKGDwWH4xw2NhZDdmN:-KgKGDwXoP2V1Qhe2Zi0:bkwiccj) ([raw file](https://github.com/robotlocomotion/drake/blob/565df173b28e825767020ce573ac6a7692e61b01/drake/automotive/prius_vis.cc#L58)):*
> ```cpp
>         std::string(parsers::FloatingJointConstants::kFloatingJointName) +
>             rpy_joint.get_position_name(i));
>   }
> ```

See my general note on this specific test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-:-KgKGrz7iWgNSpSi8c8K:butj8hf)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/prius_vis.cc, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgKFmasAVjuEbU3OV9n:-KgPdQ9H-uyTguG5-g9U:bqx3c1u) ([raw file](https://github.com/robotlocomotion/drake/blob/565df173b28e825767020ce573ac6a7692e61b01/drake/automotive/prius_vis.cc#L50)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

I would think this comment belongs in the documentation of this method's interface.  It should go somewhere in the header file.
</blockquote></details>

Since the `RigidBodyTree` is a private member variable that's initialized within this class's constructor, the only way for users to violate these requirements is to modify `/automotive/models/prius/prius_with_lidar.sdf`. Thus, instead of placing this documentation in this method's description, I think I'll place it in the class description. The unit test coverage will then be beefed up per your recommendations to detect if changes to the aforementioned SDF file breaks these requirements.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/prius_vis.cc, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgKFmasAVjuEbU3OV9n:-KgPpGSxwQo0YlP0f5d1:b-olnz1w) ([raw file](https://github.com/robotlocomotion/drake/blob/565df173b28e825767020ce573ac6a7692e61b01/drake/automotive/prius_vis.cc#L50)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Since the `RigidBodyTree` is a private member variable that's initialized within this class's constructor, the only way for users to violate these requirements is to modify `/automotive/models/prius/prius_with_lidar.sdf`. Thus, instead of placing this documentation in this method's description, I think I'll place it in the class description. The unit test coverage will then be beefed up per your recommendations to detect if changes to the aforementioned SDF file breaks these requirements.
</blockquote></details>

Sounds good to me.  I just wanted it in the header so the constraint would be clear in doxygen.  :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL. Thanks!

---

Review status: 2 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/prius_vis.cc, line 50 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgKFmasAVjuEbU3OV9n:-KgQ7hHJoopsYbXo_u7Z:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/565df173b28e825767020ce573ac6a7692e61b01/drake/automotive/prius_vis.cc#L50)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Sounds good to me.  I just wanted it in the header so the constraint would be clear in doxygen.  :)
</blockquote></details>

Done.

---

*[drake/automotive/prius_vis.cc, line 58 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgKGDwWH4xw2NhZDdmN:-KgQ7i31IPZh7aT2H4Lw:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/565df173b28e825767020ce573ac6a7692e61b01/drake/automotive/prius_vis.cc#L58)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

See my general note on this specific test.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-:-KgQ82CBdtbOnubGWw9g:bblnx7i)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Still have some concerns.

---

Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/prius_vis.cc, line 41 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgQMa7he1K_Xvhh9-EZ:-KgQMa7iHb0639K3R4QK:b-kp2h4z) ([raw file](https://github.com/robotlocomotion/drake/blob/c88c1700d5dfde6920208810ec066749f032768d/drake/automotive/prius_vis.cc#L41)):*
> ```cpp
>       dynamic_cast<const RollPitchYawFloatingJoint&>(body.getJoint());
>   DRAKE_DEMAND(rpy_joint.get_name() ==
>       parsers::FloatingJointConstants::kFloatingJointName);
> ```

BTW These last two tests confused me.

1) If the dynamic cast fails, conversion to a reference will be a run-time error. But that exception won't have the same nice wrapping as the other DRAKE_DEMANDs.
2) Testing for a joint name seems fragile.

I'd do the dynamic_cast with pointers and then DRAKE_DEMAND that the pointer isn't null. 

---

*[drake/automotive/prius_vis.h, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgQN7x_XjKRYMRzT4BB:-KgQN7x_XjKRYMRzT4BC:bnhp6s8) ([raw file](https://github.com/robotlocomotion/drake/blob/c88c1700d5dfde6920208810ec066749f032768d/drake/automotive/prius_vis.h#L20)):*
> ```objc
> /// file only contain one model instance that is not connected to the world
> /// (i.e., the SDF model must not contain a link with the same name as
> /// RigidBodyTreeConstants::kWorldName).
> ```

Is this right? If someone were to load this model multiple times into a single RBT, the method would fail. So, it's less about the contents of the file, and more about the contents of the tree.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-:-KgQOXjKPfNiHMd29rN2:b27akp6)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Concerns addressed. PTAL. Thanks again!

---

Review status: 3 of 5 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/prius_vis.cc, line 41 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgQMa7he1K_Xvhh9-EZ:-KgQUzrTNQwt129fxE-G:bs7wumb) ([raw file](https://github.com/robotlocomotion/drake/blob/c88c1700d5dfde6920208810ec066749f032768d/drake/automotive/prius_vis.cc#L41)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

BTW These last two tests confused me.

1) If the dynamic cast fails, conversion to a reference will be a run-time error. But that exception won't have the same nice wrapping as the other DRAKE_DEMANDs.
2) Testing for a joint name seems fragile.

I'd do the dynamic_cast with pointers and then DRAKE_DEMAND that the pointer isn't null. 
</blockquote></details>

Done.

---

*[drake/automotive/prius_vis.h, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgQN7x_XjKRYMRzT4BB:-KgQPxGHSZSbuVLnkDIm:b-tgikxh) ([raw file](https://github.com/robotlocomotion/drake/blob/c88c1700d5dfde6920208810ec066749f032768d/drake/automotive/prius_vis.h#L20)):*
<details><summary><i>Previously, SeanCurtis-TRI (Sean Curtis) wrote…</i></summary><blockquote>

Is this right? If someone were to load this model multiple times into a single RBT, the method would fail. So, it's less about the contents of the file, and more about the contents of the tree.
</blockquote></details>

Since this class's constructor loads the SDF file into a private tree, it's impossible for someone to load the model multiple times. The only hazard is someone changes the SDF file to prevent the model instance from being connected to the world via an RPY floating joint. This happens if someone modifies `prius_with_lidar.sdf` to contain a "world" link. (This is highly improbable, but, nonetheless, could happen.)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-:-KgQVT7WN9s983toQGOd:b-oxokyu)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
<img class="emoji" title=":LGTM:" alt=":LGTM:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks for humoring me.  :)

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/automotive/prius_vis.h, line 20 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgQN7x_XjKRYMRzT4BB:-KgQgwsr7VLzC_7Bpz6k:bkxhmxb) ([raw file](https://github.com/robotlocomotion/drake/blob/c88c1700d5dfde6920208810ec066749f032768d/drake/automotive/prius_vis.h#L20)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

Since this class's constructor loads the SDF file into a private tree, it's impossible for someone to load the model multiple times. The only hazard is someone changes the SDF file to prevent the model instance from being connected to the world via an RPY floating joint. This happens if someone modifies `prius_with_lidar.sdf` to contain a "world" link. (This is highly improbable, but, nonetheless, could happen.)
</blockquote></details>

Makes sense.  :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-:-KgQh-LGtSH0m_n3VEf3:b-5g1wvx)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Yay! +@ggould-tri  for platform review please. Thanks.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-:-KgQkZqx3bOZGOJCvON4:batj9zv)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending one discussion below.

---

Reviewed 2 of 4 files at r1, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/multibody/parsers/parser_common.cc, line 161 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgUIoqYTBClAJDxBBy2:-KgUIoqZ5QuBbcvJDxWn:btxou4k) ([raw file](https://github.com/robotlocomotion/drake/blob/ab07204a03b20433ad5563d55b6c72c76a0c4315/drake/multibody/parsers/parser_common.cc#L161)):*
> ```cpp
>     weld_to_body = tree->bodies[0].get();
>     floating_joint_name =
>         std::string(FloatingJointConstants::kFloatingJointName);
> ```

I expect that the `std::string()` explicit ctor here is redundant:  The assignment without the ctor call would be `std::string::operator=(const char*)` which AFAICT has the same semantics as `std::string::operator=(std::string(const char*))`.

Same comment for the next two instances.

(also, fyi, you already using'd std::string back at the top of the file, so you don't need the stds).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-:-KgULZYoupz2oXAx9i7F:b7ed1j7)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 4 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/multibody/parsers/parser_common.cc, line 161 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-KgUIoqYTBClAJDxBBy2:-KgURM7P10nubXIVizBQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/ab07204a03b20433ad5563d55b6c72c76a0c4315/drake/multibody/parsers/parser_common.cc#L161)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

I expect that the `std::string()` explicit ctor here is redundant:  The assignment without the ctor call would be `std::string::operator=(const char*)` which AFAICT has the same semantics as `std::string::operator=(std::string(const char*))`.

Same comment for the next two instances.

(also, fyi, you already using'd std::string back at the top of the file, so you don't need the stds).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@jwnimmer-tri for car review please. Thanks.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5638#-:-KgV8W08CxqbYiRO65PT:bdbgbfb)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
:car: good
